driver/impl/src/main/java/eu/cloudnetservice/driver/impl/template/RemoteTemplateStorage.java

```diff
@@ -140,22 +170,63 @@ public boolean deploy(@NonNull ServiceTemplate target, @NonNull InputStream inpu
    * {@inheritDoc}
    */
   @Override
-  public @NonNull OutputStream appendOutputStream(
+  public @NonNull CompletableFuture<ZipInputStream> openZipInputStreamAsync(@NonNull ServiceTemplate template) {
+    return this.zipTemplateAsync(template).thenApply(inputStream -> inputStream != null
+      ? new ZipInputStream(inputStream)
+      : null);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public @NonNull CompletableFuture<OutputStream> appendOutputStreamAsync(
     @NonNull ServiceTemplate template,
     @NonNull String path
-  ) throws IOException {
-    return this.openLocalOutputStream(template, path, FileUtil.createTempFile(), true);
+  ) {
+    return TaskUtil.supplyAsync(() -> this.appendOutputStream(template, path));
   }
 
   /**
    * {@inheritDoc}
    */
   @Override
-  public @NonNull OutputStream newOutputStream(
+  public @NonNull CompletableFuture<OutputStream> newOutputStreamAsync(
     @NonNull ServiceTemplate template,
     @NonNull String path
-  ) throws IOException {
-    return this.openLocalOutputStream(template, path, FileUtil.createTempFile(), false);
+  ) {
+    return TaskUtil.supplyAsync(() -> this.newOutputStream(template, path));
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public @NonNull CompletableFuture<Boolean> deployDirectoryAsync(
+    @NonNull ServiceTemplate target,
+    @NonNull Path directory,
+    @Nullable Predicate<Path> filter
+  ) {
+    try (var inputStream = ZipUtil.zipToStream(directory, filter)) {
+      return this.deployAsync(target, inputStream);
+    } catch (IOException exception) {
+      return CompletableFuture.completedFuture(false);
```

Comment:
Shouldn't this return a failed future with the exception?

Reply:
I guess the zipToStream should be in a supplyAsync call that composes the deploy step afterwards? something like:
```java
return CompletableFuture.supplyAsync(() -> ZipUtil.zipToStream(directory, filter))
  .thenCompose(inputStream -> this.deployAsync(target, inputStream));
```